inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/InstrumentationUpdateService.java

```diff
@@ -0,0 +1,176 @@
+package rocks.inspectit.oce.core.instrumentation;
+
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import net.bytebuddy.description.type.TypeDescription;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Service;
+import rocks.inspectit.oce.core.config.InspectitConfigChangedEvent;
+import rocks.inspectit.oce.core.config.InspectitEnvironment;
+import rocks.inspectit.oce.core.config.model.instrumentation.InstrumentationSettings;
+import rocks.inspectit.oce.core.config.model.instrumentation.InternalSettings;
+import rocks.inspectit.oce.core.instrumentation.special.SpecialSensor;
+import rocks.inspectit.oce.core.selfmonitoring.SelfMonitoringService;
+import rocks.inspectit.oce.core.service.BatchJobExecutorService;
+import rocks.inspectit.oce.core.utils.CommonUtils;
+import rocks.inspectit.oce.core.utils.StopWatch;
+
+import javax.annotation.PostConstruct;
+import java.lang.instrument.Instrumentation;
+import java.time.Duration;
+import java.util.*;
+
+/**
+ * This class is responsible for making sure that for every class the instrumentation
+ * eventually is applied as specified in the active {@link rocks.inspectit.oce.core.config.model.InspectitConfig}
+ * <p>
+ * Whenever a classes instrumentation configuration might have changes, this class gets put into this services working queue.
+ * This service is then responsible for filtering out the classes whose instrumentaiton actually has
+ * changed and submits them to the {@link ClassRetransformationService}.
+ */
+@Service
+@Slf4j
+public class InstrumentationUpdateService {
+
+    @Autowired
+    private BatchJobExecutorService executor;
+
+    @Autowired
+    private InspectitEnvironment env;
+
+    @Autowired
+    private List<SpecialSensor> specialSensors;
+
+    @Autowired
+    private SelfMonitoringService selfMonitoring;
+
+    @Autowired
+    private AsyncClassTransformer transformer;
```

Comment:
if this class implements `IPostInstrumentationListener` and `IClassesDiscoveryListener` then you don't need these direct dependences.. Also most of the stuff in the `init()` can be removed..

Reply:
The direct dependency to the transformer is still required to detect if it is shutting down.
Otherwise we constantly issue classes for retransform even though they will never be instrumented..